### ISSUE TITLE
Upper limit on zip file size?

### ISSUE THREAD

**Issue by gannons** [Stuk/jszip#580]
I've been asked if it's possible to generate large zips in the browser. From my research, and experiments, it's starting to look like this is not what the platform is designed for. AFAIK if we attempt to generate large files (>= ~3 Gb) in the browser we'll run into memory issues / crashes. But I'm having difficultly confirming this with specifications, documentation, wikis, etc.

I was hoping the contributors might have an opinion on what is a reasonable upper limit? i.e. For files generated by jszip?

**Comment by florianstancioiu** [comment]
I'm wondering about this limitation too.
Can we get an answer, some people over on reddit talk about a 5mb size limit, is that true?

**Comment by sregger** [comment]
> on reddit talk about a 5mb size limit, is that true?

No. Perhaps it was true in the past but modern browsers can generate much larger zip files. In my tests I managed to generate a 10-20 GB zip.

However the browse does start to become unstable and will eventually crash. I was also using generated, as opposed to real, data so your mileage may vary.

In my scenario we moved away from generating the zip in a browse and settled on a dedicated desktop client which we wrote with Electron.

**Comment by fuscoem** [comment]
Using version 3.2.2 of jszip and streamsaver we managed to generate zip files in streaming in a vue app (in browser) using the generateNodeStream. Unfortunately with newer versions of jszip the generateNodeStream does not work in a browser (see https://github.com/Stuk/jszip/issues/663 )

**Comment by bybraveHQ** [comment]
An update for people finding this thread: the practical answer to "what's the upper limit" in stock JSZip is 4 GB — not because of the browser, but because JSZip never wrote ZIP64 structures, so any size, offset or entry count beyond the classic 32/16-bit fields produced a corrupted archive (silently, in most paths).

I maintain an actively developed fork ([@bybrave/jszip2](https://github.com/bybraveHQ/jszip2)); since 4.3.0 it writes ZIP64 automatically — when a size, an offset or the entries count overflows, the affected entries get their ZIP64 extra fields and the archive gets a ZIP64 end-of-central-directory record. Nothing to configure:

```js
const JSZip = require("@bybrave/jszip2"); // drop-in fork of jszip

zip.file("huge.bin", hugeStream);                 // node or web stream
zip.generateNodeStream().pipe(output);            // > 4 GB? ZIP64 kicks in by itself
```

On the memory side (the other half of this issue): with streamed input and streamed output the content is never held in memory at once — the fork also has web streams (`generateWebStream()`, 4.2.0), so browser-side you can pipe straight to the File System Access API instead of materializing a Blob. The only case needing a flag is `streamFiles: true` with >4 GiB entries (`zip64: true`, explained by an explicit error).

Archives with >4 GB files and 65 540 entries verified with Info-ZIP `unzip -t`. Release notes: https://github.com/bybraveHQ/jszip2/releases/tag/v4.3.0
